### ISSUE TITLE
nested math in token elements.

### ISSUE THREAD

**Issue by davidcarlisle** [w3c/mathml#522]
The following issue arose while working on examples using MathML for accessibility tagging in PDF, looking at typical tex cases
such as a display containing nested text such as

```
\[
f(n) = \begin{cases}
        1 & \text{if $n$ is prime} \\
        0 & \text{otherwise}
       \end{cases}
\]
```

$$
f(n) =\begin{cases}
    1 & \text{if $n$ is prime} \\
     0 & \text{otherwise}
    \end{cases}
$$

MathML3 and 4 allow HTML elements in presentation token elements when used in html, deferring to the HTML(5) spec for exactly which elements are allowed.

HTML specifies the phrasing content is allowed 

https://html.spec.whatwg.org/#phrasing-content-2

The schema used with mathml3 and the current draft mathml-core schema interpreted this by extending the content model of token elements to take text or html, this means that to make a valid version of the above with a nested `$n$`
 in `<mtext>` you need an intervening html `<span>`

```
<mtext><span>if <math><mi>n</mi></math> is prime</span></mtext>
```

However it was pointed out, correctly I think, that phrasing content at the above link, explicitly lists mathml namespace `<math>` and svg  namespace `<svg>` as phrasing content, so that they should be valid children of `<mtext>` and that current browsers do all implement that correctly and this mathml does render as intended.

```
<mtext>if <math><mi>n</mi></math> is prime</mtext>
```

This could be fixed in the draft core schema by changing the `textorHTML` production

https://w3c.github.io/mathml/#parsing_textorHTML

to include `| math | svg`

and possibly to highlight this in the section on mathml in html at 

https://w3c.github.io/mathml/#interf_html

Given that HTML has always specified phrasing content this way and that it is implemented this way in browsers, I think changing the schema is essentially an editorial correction (especially as only draft versions of it have been published so far anyway)

Apart from allowing simpler markup and matching current browser implementations, allowing math directly under mtext means that the common case of nested math which _just_ has text and math such as `\text{when $x<0$}` can be modelled just using a single, mathml, namespace. You only need to bring in the additional html namespace to model nested textual markup such as `\text{something \textbf{bold} about $x$}` which is far less common.

So while arguing that this could be viewed as an editiorial correction I think it's a big enough change that it should be logged as an issue, hence this issue.

**Comment by physikerwelt** [comment]
Does this only apply to mtext, or also to mi, mo, mn, ms? In https://phabricator.wikimedia.org/T389712, we identified those elements as leaf elements in the tree. After reading https://w3c.github.io/mathml-core/ we got the impression that those elements would usually have only text children.

**Comment by davidcarlisle** [comment]
@physikerwelt  In practice it is mtext where this is needed but mathml (and html) have always treated all mathml token elements as having the same content model. 

"pure" mathml has always allowed text or mglyph but this proposal suggests allowing math as well.

Mathml-in-html allows text or html flow content (so inline html elements including math)

It's true the wording in mathml-core is a bit different for mtext than the other token elements, but in practice they all need to allow inline html flow elements (not least to allow html `<a>` links as native mathml linking is not supported)

**Comment by physikerwelt** [comment]
So that means one could use `<mi><a href="https://en.wikipedia.org/wiki/Planck_constant">ℏ</a></mi>` instead of `<a href="https://en.wikipedia.org/wiki/Planck_constant"><mi>ℏ</mi></a>` which we currently use. 

I would image `\text{something \textbf{bold} about $x$}` turns into <mrow><mtext>someting</mtext>... ? 

temml.org produces

`<mrow><mtext>something&nbsp;𝐛𝐨𝐥𝐝&nbsp;about&nbsp;</mtext><mi>x</mi></mrow>`

mathoid does not support commands inside text

[LaTeXML](https://latexml.mathweb.org/editor) generates full HTML output

`<p id="p1.1" class="ltx_p"><span id="p1.1.m1.1.1.1" class="ltx_text ltx_markedasmath">something <span id="p1.1.m1.1.1.1.1" class="ltx_text ltx_font_bold">bold</span> about <math id="p1.1.m1.1.1.1.m1.1" class="ltx_Math" alttext="x" display="inline"><semantics id="p1.1.m1.1.1.1.m1.1a"><mi id="p1.1.m1.1.1.1.m1.1.1" xref="p1.1.m1.1.1.1.m1.1.1.cmml">x</mi><annotation-xml encoding="MathML-Content" id="p1.1.m1.1.1.1.m1.1b"><ci id="p1.1.m1.1.1.1.m1.1.1.cmml" xref="p1.1.m1.1.1.1.m1.1.1">𝑥</ci></annotation-xml><annotation encoding="application/x-tex" id="p1.1.m1.1.1.1.m1.1c">x</annotation><annotation encoding="application/x-llamapun" id="p1.1.m1.1.1.1.m1.1d">italic_x</annotation></semantics></math></span></p>`

MathJax generates
```
<math xmlns="http://www.w3.org/1998/Math/MathML" data-latex="\text{something \textbf{bold} about $x$}" display="block">
  <mrow data-latex="\text{something \textbf{bold} about $x$}">
    <mtext>something \textbf{bold} about&#xA0;</mtext>
    <mrow data-mjx-texclass="ORD">
      <mi data-latex="x">x</mi>
    </mrow>
  </mrow>
</math>
```
So my feeling is that either using an outer mrow or wrapping different elements into an inner span are already sufficient options.

**Comment by davidcarlisle** [comment]
On Thu, 26 Jun 2025 at 22:47, Moritz Schubotz ***@***.***>
wrote:

> *physikerwelt* left a comment (w3c/mathml#522)
> <https://github.com/w3c/mathml/issues/522#issuecomment-3010242830>
>
> So that means one could use <mi><a href="
> https://en.wikipedia.org/wiki/Planck_constant">ℏ</a></mi> instead of <a
> href="https://en.wikipedia.org/wiki/Planck_constant"><mi>ℏ</mi></a> which
> we currently use.
>

that is already the case. The first form with a inside mi  is valid HTML5,
the second is invalid (although it may possibly work in some browsers)  as
html <a element is only allowed in token elements not as a child of other
mathml elements.

> I would image \text{something \textbf{bold} about $x$} turns into
> someting... ?
>
> temml.org produces
>
>
> <mrow><mtext>something&nbsp;𝐛𝐨𝐥𝐝&nbsp;about&nbsp;</mtext><mi>x</mi></mrow>
>
> mathoid does not support commands inside text
>
yes in simple cases you can flatten out the nesting but it's not always so
easy.

> LaTeXML <https://latexml.mathweb.org/editor> generates full HTML output
>
> <p id="p1.1" class="ltx_p"><span id="p1.1.m1.1.1.1" class="ltx_text
> ltx_markedasmath">something <span id="p1.1.m1.1.1.1.1" class="ltx_text
> ltx_font_bold">bold</span> about <math id="p1.1.m1.1.1.1.m1.1"
> class="ltx_Math" alttext="x" display="inline"><semantics
> id="p1.1.m1.1.1.1.m1.1a"><mi id="p1.1.m1.1.1.1.m1.1.1"
> xref="p1.1.m1.1.1.1.m1.1.1.cmml">x</mi><annotation-xml
> encoding="MathML-Content" id="p1.1.m1.1.1.1.m1.1b"><ci
> id="p1.1.m1.1.1.1.m1.1.1.cmml"
> xref="p1.1.m1.1.1.1.m1.1.1">𝑥</ci></annotation-xml><annotation
> encoding="application/x-tex"
> id="p1.1.m1.1.1.1.m1.1c">x</annotation><annotation
> encoding="application/x-llamapun"
> id="p1.1.m1.1.1.1.m1.1d">italic_x</annotation></semantics></math></span></p>
>
> MathJax generates
>
> <math xmlns="http://www.w3.org/1998/Math/MathML" data-latex="\text{something \textbf{bold} about $x$}" display="block">
>   <mrow data-latex="\text{something \textbf{bold} about $x$}">
>     <mtext>something \textbf{bold} about&#xA0;</mtext>
>     <mrow data-mjx-texclass="ORD">
>       <mi data-latex="x">x</mi>
>     </mrow>
>   </mrow>
> </math>
>
> So my feeling is that either using an outer mrow or wrapping different
> elements into an inner span are already sufficient options.
>

the mathjax output you show  is using \textbf for the nested bold so not
really the final xml/html form.

The latexml output simplified is

<p>
<span>something <span>bold</span> about <math>
<mi>x</mi>
</math>
</span>
</p>

which is fine if the original `\text` was not in math mode so it can all be
text apart from the single x but this mostly comes up in things like
\begin{cases}... where the whole thing is deeply embedded in a math display
so you can not introduce a top level p

David

**Comment by dginev** [comment]
re: @physikerwelt 
> So my feeling is that either using an outer mrow or wrapping different elements into an inner span are already sufficient options.

---

The thing is, `<math>` will be allowed in MathML Core, wherever HTML content is allowed. As David quoted in the issue description:

> However it was pointed out, correctly I think, that phrasing content at the above link, explicitly lists mathml namespace <math> and svg namespace <svg> as phrasing content

I think that is healthy, especially for `<mtext>`. And it avoids special-casing at the implementation level. Implementers certainly don't have to use the capability if it isn't needed.

Since MathML Full has taken the position that it should be a superset of MathML Core, making it more restrictive doesn't quite fit the philosophy... So I'm in favor of the issue as posed.

**Comment by MurrayIII** [comment]
I added some simple code to my LaTeX input converter (see code for "Flatten embedded math zones" in [TeX.js](https://github.com/MurrayIII/UnicodeMathML/blob/main/playground/assets/TeX.js)) that flattens the tree in the initial example. Essentially the first $ terminates the mtext and the math zone continues with the 𝑛. It could be 𝑛², etc. Then a second mtext handles what follows the second $. So you can copy the sample LaTeX above and paste it into the input field of [UnicodeMathML](https://murrayiii.github.io/UnicodeMathML/playground/) and see

<img width="1305" height="566" alt="Image" src="https://github.com/user-attachments/assets/e5314a8d-7cd5-4e8a-9592-379bbbf1aafe" />

**Comment by physikerwelt** [comment]
For search we are currently using LaTeXML to generate queries like
```xml
<?xml version="1.0"?>
<mws:query xmlns:mws="http://search.mathweb.org/ns" xmlns:m="http://www.w3.org/1998/Math/MathML" limitmin="0" answsize="30000" totalreq="yes">
  <mws:expr>
    <m:apply xml:id="p1.1.m1.1.3.cmml" xref="p1.1.m1.1.3">
      <m:sin xml:id="p1.1.m1.1.1.cmml" xref="p1.1.m1.1.1"/>
      <mws:qvar>x</mws:qvar>
    </m:apply>
  </mws:expr>
</mws:query>
```
I was planning to upgrade this to combine mathml-core instead of http://www.w3.org/1998/Math/MathML

If we accepted, would this mean I can also use mws:qvar within m:mtext? That would be actually quite handy.

By the way, didn't we vote in one of the last MathML meetings to resolve this issue, by following the suggestions and discouraging the use of nested MathML at the same time.

**Comment by dginev** [comment]
@physikerwelt I think that snippet you've quoted uses a Content MathML search engine, with the qvar extension for latexml [here](https://github.com/MathWebSearch/LaTeXML-Plugin-MathWebSearch/blob/master/lib/LaTeXML/Package/mws.sty.ltxml). 

From what I remember, that search engine never relied on a validation step for the MathML, and could always index invalid trees, as long as they were well-formed. In any case, I think that is more a question for the MathWebSearch team, than for the MathML spec.

**Comment by physikerwelt** [comment]
@dginev maybe that example is too specific. I mean, can I just put HTML inside the mtext element, or any XML that is allowed by the root document

**Comment by davidcarlisle** [comment]
@physikerwelt 

> can I just put HTML inside the mtext element

Since HTML5 you have always been able to put any html phrasing content elements (ie inline content, essentially) in mtext. This PR does not change that, it just confirms (what the HTML spec says) that math and svg are HTML phrasing content 
elements. https://html.spec.whatwg.org/#phrasing-content-2 so yes you can use html in mtext in mathml-in-html.
You can (therefore) use math in mtext in mathml-in-html, this PR is proposing that math in mtext is allowed in "pure" mathml without needing to reference the HTML specification to allow it.

**Comment by physikerwelt** [comment]
@davidcarlisle sorry for my stupid questions, but I have the feeling that I still don't understand the reasoning.

I can use MathML in many contexts, HTML, MathSeach, and TEI... So the HTML spec is only one (important) special case, correct?

The MathML spec defines
```rnc
mtext = element mtext {mtext.attributes, token.content}

token.content = textorHTML

textorHTML = text | element (h:*)  {attribute * {text}*,textorHTML*}

namespace h = "http://www.w3.org/1999/xhtml"
```
How should that be changed?

For [TEI](https://www.tei-c.org/release/xml/tei/custom/schema/relaxng/tei_math.rng) it seems that they explicitly overwrite the spec and don't allow anything within mtext
```xml
      <define name="mathml.mtext">
         <element name="mtext">
            <ref name="mathml.attlist-mtext"/>
            <zeroOrMore>
               <choice>
                  <ref name="mathml.MathMLCharacters"/>
                  <ref name="mathml.malignmark"/>
               </choice>
            </zeroOrMore>
         </element>
      </define>
```

This is a bit unfortunate for publishing workflows, when TeX is converted to TEI, and for example, Links should be used. Since most people won't validate their XML files, this might not be a practical problem. However, I would be willing to understand what is theoretically allowed by the spec and what is not.

**Comment by davidcarlisle** [comment]
@physikerwelt  Mathml has always specified that mtext and other token elements  just take text or mglyph content, but  has (since mathml 3) stated in words that vocabularies hosting mathml can extend mathml at these points to allow additional content.

The specific example of mathml-in-html is given which allows html in mtext.

"pure" mathml can not allow this as it has no way to refer to html (or TEI or docbook or anything else embedding mathml)
This PR notes that the one case that could/should be allowed in general is math as that does not involve any extended vocabulary.

So TEI does not _restrict_ mtext, it just chooses not to extend it it could extend it by adjusting the mtext content model to allow inline TEI elements.

**Comment by physikerwelt** [comment]
> This PR notes that the one case that could/should be allowed in general is math as that does not involve any extended vocabulary.

I think I understand that, and don't find that a good idea. I don't see any reason why math should be allowed, while, for example, SVG would not be allowed. I think it should not be the job of the MathML spec to define that, but rather to the spec that embeds MathML elements, how the "text" elements are extended.

**Comment by davidcarlisle** [comment]
@physikerwelt  the situation with math and svg are completely different. Any vocabulary using mathml has a defined math element, that is not the case for svg. If every practical application is extending mathml to allow math it makes sense to allow math in the base specification.  One practical example is PDF tagging (which only allows mathml)

**Comment by physikerwelt** [comment]
>  One practical example is PDF tagging (which only allows mathml)

Is this really that useful? I mean, one would like to put any tags that apply to the elements that are used within the MathML text element. So I think one has to define it on a case-by-case basis. On the other side, maybe one does not want to allow math (e.g., to reduce complexity).

**Comment by davidcarlisle** [comment]
> > One practical example is PDF tagging (which only allows mathml)
> 
> Is this really that useful? I mean, one would like to put any tags that apply to the elements that are used within the MathML text element. So I think one has to define it on a case-by-case basis. On the other side, maybe one does not want to allow math (e.g., to reduce complexity).

useful or not the set of element names is given (by the ISO pdf standard) and Includes math but does not include svg.

but the main motivation here is not pdf, it is that math is the only element defined in the mathml schema that makes sense at this point, the base schema can not allow html or svg or TEI elements they have to be an extension related to the embedding in those vocabularies.

The current proposal (from a recent zoom) is to make the base schema just allow text as now but have the extension allowing math as a kind of "example extension" I took an action (not yet done) to make a PR with specific text changes for the spec and schemas.

**Comment by physikerwelt** [comment]
Yes, I see that, but the big downside is that one can no longer opt-out of having math within text (like the current MathJax implementation does).

**Comment by davidcarlisle** [comment]
@physikerwelt  I think that opt out will be necessary, and anyway couldn't be prevented. If mtext is an extension point where systems can define the content model they need, then "just text" must be one possibility that could be defined.

But as I say I think the current plan is that "just text" will be the default with allow math as a "provided extension" with exact details to be worked out.

**Comment by NSoiffer** [comment]
FYI: MathJax 4 was just released and one of the new features is "Support for HTML embedded in TeX and MathML expressions".

**Comment by physikerwelt** [comment]
Yeah, that’s what one would reasonably expect. What is proposed here is support MathML within MathML. I would rather want to limit alternatives to express the same thing. In the same way we could also support links with a elements within tokens.

**Comment by davidcarlisle** [comment]
> Yeah, that’s what one would reasonably expect. What is proposed here is support MathML within MathML. I would rather want to limit alternatives to express the same thing. In the same way we could also support links with a elements within tokens.

yes but html-in-token-elements implies mathml (and svg) in token elements as they are included in the relevant html, so the question is whether to directly specify that or only allow it when you are allowing html.

**Comment by physikerwelt** [comment]
The question is if MathML 4 (core) should be usable outside of something like HTML. My feeling has always been that we more and more rely on CSS and other web standards and thus MathML always has to be embedded into something. I think for MathML authors it would be good to have a description of this something. Even if we won't be able to fully formally define that in the relax-ng schema, I guess we can get close.

**Comment by physikerwelt** [comment]
Maybe something like
```
mtext = element mtext {mtext.attributes, token.content}

token.content = textorContext

textorContext= text | element (h:*)  {attribute * {text}*,textorContext*}

namespace h = // this needs to become an external variable
```
In some virtual postprocessing step the XML tree an be transformed to HTML (without ns)

**Comment by dginev** [comment]
There is a beautiful rnc trick used by SVG 1.1 which allows extensibility "in spirit". It is:

```
## 
##       any element in foreign namespace
##     
foreignElement =
    element * - svg:* {
        (attribute * - svg:* { text }
         | text
         | foreignElement)*
    }
```

And if you look at that closely enough, you'll notice that the `<math>` used in HTML is still in the MathML namespace, even if it enters through the flow content rule of HTML. So as David says, even if we put some extra effort to "announce" foreign element extensions in the schema, we would continue to need a `<math>`-specific rule allowing the one exceptional extension point that is still in the same namespace.

Edit: Maybe that is a healthy compromise solution - introduce an "Extensibility" class in the schema that only contains `math` by default, and allow stricter uses of MathML to redefine it to `notAllowed`. Or define that ourselves as some "strict" mode...

**Comment by physikerwelt** [comment]
> So as David says, even if we put some extra effort to "announce" foreign element extensions in the schema, we would continue to need a <math>-specific rule allowing the one exceptional extension point that is still in the same namespace.

Ok, so it's rather a technical detail. To make math within math nothing special or noteworthy. And we are not saying you need to be able to parse math within math. My goal is the neither the inner math nor the outer MathML element needs to be aware of the other.

**Comment by davidcarlisle** [comment]
@dginev 

> There is a beautiful rnc trick used by SVG 1.1 which allows extensibility "in spirit". It is:

yes the mathml schema is essentially the same. (I wonder where they got that from:-)

As in mathml3 we allow any attribute not in mathml or no-namespce

`NonMathMLAtt = attribute (* - (local:* | m:*)) {xsd:string}`

and mtext (in its mathml core version) allows any html

```
textorHTML = text | element (h:*)  {attribute * {text}*,textorHTML*}
			    
token.content = textorHTML
```

here `h:*` was supposed to be an over-estimate of "HTML phrasing content" (to save importing the entire HTML spec) but the main content of this issue is that simple form `h:*` misses two important elements that are listed as phrasing content in the HTML spec, namely `math` and `svg` . So essentially this issue is just asking for a technical fix to the RelaxNG and not any substantive change to the spec.

**Comment by davidcarlisle** [comment]
I will  adjust the mathml4 text to allow this (mathml zoom)

**Comment by davidcarlisle** [comment]
closing this as mathml4 spec and schema both updated.
